# 声明式事务-基于XML
## 一、实现原理
基于AOP实现的，Advice为 `org.springframework.transaction.interceptor.TransactionInterceptor`。 `TransactionInterceptor` 是由 `TxAdviceBeanDefinitionParser` 解析成 `BeanDefinition` 而后再创建的。
### TxAdviceBeanDefinitionParser
将 `<tx:advice>` 解析成 BeanDefinition 并依此创建 Bean。 参考 [BeanDefinitionParser](../4、spring-beans/8.%20BeanDefinitionParser.md)

```java
class TxAdviceBeanDefinitionParser extends AbstractSingleBeanDefinitionParser {
    @Override
    protected Class<?> getBeanClass(Element element) {
        return TransactionInterceptor.class;
    }
    
    // ------------------
    // other code ignored
    // ------------------
}
```

### TransactionInterceptor逻辑
核心逻辑位于 `TransactionInterceptor` 的父类 `TransactionAspectSupport` 中。
```java
public abstract class TransactionAspectSupport implements BeanFactoryAware, InitializingBean {
    protected Object invokeWithinTransaction(Method method, @Nullable Class<?> targetClass,
                                             final InvocationCallback invocation) throws Throwable {
        // ------------------
        // other code ignored
        // ------------------
        try {
            // 执行原方法
            retVal = invocation.proceedWithInvocation();
        } catch (Throwable ex) {
            // 有异常就回滚
            completeTransactionAfterThrowing(txInfo, ex);
            throw ex;
        } finally {
            // 没异常就提交
            cleanupTransactionInfo(txInfo);
        }

        commitTransactionAfterReturning(txInfo);
        return retVal;
    }
    // ------------------
    // other code ignored
    // ------------------
}
```
简单来说就是使用Around切面，在原方法的执行上加上try-catch-finally，如果有异常就回滚，没有异常就提交。
#### 回滚条件
虽说在 `TransactionAspectSupport` 里如果有异常就会执行 `completeTransactionAfterThrowing`，但并非所有情况都会回滚。其逻辑如下：
```java
public abstract class TransactionAspectSupport implements BeanFactoryAware, InitializingBean {
    protected void completeTransactionAfterThrowing(@Nullable TransactionInfo txInfo, Throwable ex) {
        if (txInfo != null && txInfo.getTransactionStatus() != null) {
            // ------------------
            // code simplified
            // ------------------
            if (txInfo.transactionAttribute != null && txInfo.transactionAttribute.rollbackOn(ex)) {
                txInfo.getTransactionManager().rollback(txInfo.getTransactionStatus());
            } else {
                txInfo.getTransactionManager().commit(txInfo.getTransactionStatus());
            }
        }
        // ------------------
        // other code ignored
        // ------------------
    }
}
```
也就是具体看 `txInfo.transactionAttribute.rollbackOn(ex)` 的计算结果会不会回滚。逻辑如下：
```java
public class RuleBasedTransactionAttribute extends DefaultTransactionAttribute implements Serializable {
    @Override
    public boolean rollbackOn(Throwable ex) {
        RollbackRuleAttribute winner = null;
        int deepest = Integer.MAX_VALUE;

        if (this.rollbackRules != null) {
            for (RollbackRuleAttribute rule : this.rollbackRules) {
                int depth = rule.getDepth(ex);
                if (depth >= 0 && depth < deepest) {
                    deepest = depth;
                    winner = rule;
                }
            }
        }

        // User superclass behavior (rollback on unchecked) if no rule matches.
        if (winner == null) {
            return super.rollbackOn(ex);
        }

        return !(winner instanceof NoRollbackRuleAttribute);
    }
}
```
也就是看 `rollback-for` 指定的异常（还可以通过depth检查inner Exception）。在没有指定 `rollback-for` 的情况下，**默认是 `RuntimeException` 或 `Error` 才会回滚。**
```java
public class DefaultTransactionAttribute extends DefaultTransactionDefinition implements TransactionAttribute {
    @Override
    public boolean rollbackOn(Throwable ex) {
        return (ex instanceof RuntimeException || ex instanceof Error);
    }   
}
```
A pattern can be a fully qualified class name or a substring of a fully qualified class name for an exception type (which must be a subclass of Throwable), with no wildcard support at present. For example, a value of "javax.servlet.ServletException" or "ServletException" will match javax.servlet.ServletException and its subclasses.
当一个回滚规则被定义为一个exception类型时，该类型将被用来与抛出的exception及其超类型相匹配，提供了类型安全并避免了使用模式时可能发生的任何意外匹配。例如，jakarta.servlet.ServletException.class 的值将只匹配 jakarta.servlet.ServletException 类型及其子类的抛出的exception。

Consequently, @Transactional(rollbackFor = example.CustomException.class) is equivalent to @Transactional(rollbackForClassName = "example.CustomException").

For example, given a rule configured to match on com.example.CustomException, that rule would match against an exception named com.example.CustomExceptionV2 (an exception in the same package as CustomException but with an additional suffix) or an exception named com.example.CustomException$AnotherException (an exception declared as a nested class in CustomException).
#### 手动指定rollback-for
```xml
<tx:advice id="txAdvice" transaction-manager="txManager">
    <tx:attributes>
        <tx:method name="get*" read-only="true"/>
        <tx:method name="*" rollback-for="java.lang.Exception"/>
    </tx:attributes>
</tx:advice>
```
